src/dithering_demo/main.cpy

```diff
@@ -0,0 +1,224 @@
+
+class ImageDropdown : public ui::TextDropdown:
+  public:
+  IApp *app
+  vector<string> filenames
+
+  ImageDropdown(int x, y, w, h, IApp *app) \
+      : app(app), \
+      ui::TextDropdown(x, y, w, h, "Image selection"):
+    self.dir = ui::DropdownButton::DIRECTION::DOWN
+    self.populate()
+
+  void populate():
+    section := add_section("Select an image");
+    self.filenames = read_directory(IMAGES_DIR)
+    section->add_options(self.filenames)
+
+  void on_select(int idx):
+    app->on_image_selected(self.filenames[idx])
+
+#define WAVEFORM_PAIR(X) std::make_pair<>( #X, X )
+
+class WaveformDropdown : public ui::TextDropdown:
+  public:
+  IApp *app
+  vector<pair<string,int>> modes = { \
+     WAVEFORM_PAIR(WAVEFORM_MODE_DU), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_GC16), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_GC4), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_A2), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_DU4), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_AUTO) \
+  }
+
+  WaveformDropdown(int x, y, w, h, IApp *app) \
+      : app(app), \
+      ui::TextDropdown(x, y, w, h, "Waveform selection"):
+    self.dir = ui::DropdownButton::DIRECTION::DOWN
+    self.populate()
+
+  void populate():
+    section := add_section("Waveform mode")
+    vector<string> labels
+    for (auto mode : self.modes)
+      labels.push_back(mode.first)
+    section->add_options(labels)
+
+  void on_select(int idx):
+    app->on_waveform_selected(modes[idx].second)
+
+#define DITHER_PAIR(X) std::make_pair<>( #X, framebuffer::DITHER::X )
+
+class DitheringDropdown : public ui::TextDropdown:
+  public:
+  IApp *app
+  vector<pair<string,framebuffer::DITHER::MODE>> modes = { \
+     DITHER_PAIR(NONE), \
+     DITHER_PAIR(BAYER_2), \
+     DITHER_PAIR(BAYER_16), \
+     DITHER_PAIR(BLUE_NOISE_2), \
+     DITHER_PAIR(BLUE_NOISE_16), \
+  }
+
+  DitheringDropdown(int x, y, w, h, IApp *app) \
+      : app(app), \
+      ui::TextDropdown(x, y, w, h, "Dithering selection"):
+    self.dir = ui::DropdownButton::DIRECTION::DOWN
+    self.populate()
+
+  void populate():
+    section := add_section("Dithering mode")
+    vector<string> labels
+    for (auto mode : self.modes)
+      labels.push_back(mode.first)
+    section->add_options(labels)
+
+  void on_select(int idx):
+    app->on_dithering_selected(modes[idx].second)
+
+class App : public IApp:
+  public:
+
+  ImageDropdown * image_selector;
+  WaveformDropdown * waveform_selector;
+  DitheringDropdown * dithering_selector;
+
+  ui::Text * undithered_label
+  DitheredBitmap *undithered_bmp
+  ui::Text * dithered_label
+  DitheredBitmap *dithered_bmp
+
+  image_data image = { NULL, 0 }
+  int waveform_mode = WAVEFORM_MODE_DU
+
+  App():
+    demo_scene := ui::make_scene()
+    ui::MainLoop::set_scene(demo_scene)
+
+    ui::Button::DEFAULT_STYLE += ui::Stylesheet().border_all().justify_center().valign_middle()
+
+    fb := framebuffer::get()
+    fb->clear_screen()
+    fb->redraw_screen()
+    int w, h
+    w, h = fb->get_display_size()
```

Comment:
`w, h := fb->get_display_size(); ` is also fine, iirc

Reply:
👍 happy to change. I had that initially and ended up with a weird error message that suggested I was calling `new DitheredBitmap(int, int, int&, int)`, but I don't see that now (I think it was before I added the explicit constructor).